File: Adapter Design Pattern/C++ Code/AdapterDesignPattern.cpp
#include <iostream>

// Duck interface
class Duck {
public:
    virtual void quack() = 0;
    virtual void fly() = 0;
};

// Turkey interface
class Turkey {
public:
    virtual void gobble() = 0;
    virtual void fly() = 0;
};

// Concrete duck implementation
class MallardDuck : public Duck {
public:
    void quack() override {
        std::cout << "Quack!" << std::endl;
    }
    void fly() override {
        std::cout << "I'm flying!" << std::endl;
    }
};

// Concrete turkey implementation
class WildTurkey : public Turkey {
public:
    void gobble() override {
        std::cout << "Gobble gobble!" << std::endl;
    }
    void fly() override {
        std::cout << "I'm flying a short distance!" << std::endl;
    }
};

// Turkey adapter using multiple inheritance
class TurkeyAdapter : public Duck, public WildTurkey {
public:
    void quack() override {
        gobble();
    }
    void fly() override {
        for (int i = 0; i < 5; i++) {
            WildTurkey::fly();
        }
    }
};

// Client code
int main() {
    MallardDuck* duck = new MallardDuck();
    WildTurkey* turkey = new WildTurkey();
    Duck* turkeyAdapter = new TurkeyAdapter();

    std::cout << "The turkey says..." << std::endl;
    turkey->gobble();
    turkey->fly();

    std::cout << "\nThe duck says..." << std::endl;
    duck->quack();
    duck->fly();

    std::cout << "\nThe turkey adapter says..." << std::endl;
    turkeyAdapter->quack();
    turkeyAdapter->fly();

    delete duck;
    delete turkey;
    delete turkeyAdapter;

    return 0;
}

File: Adapter Design Pattern/Adapter Design Pattern Java Code/AdapterDesigPattern/src/AdapterDesignPattern.java
interface Duck{
    void Quack();
    void Fly();
}
interface Turkey{
    void Gobble();
    void Fly();
}
class MallardDuck implements Duck {
    public void Quack(){
        System.out.println("Duck Doing Quack Quack");
    }
    public void Fly(){
        System.out.println("Duck Doing Fly Fly");
    }
}
class WildTurkey implements Turkey{
    public void Gobble(){
        System.out.println("Turkey Doing Gobble Gobble");
    }
    public void Fly(){
        System.out.println("Turkey Doing Fly Fly");
    }
}
class AdapterForTurkey implements Duck {
    Turkey Trk; //Object Adapter
    AdapterForTurkey(Turkey T1){
        this.Trk=T1;
    }
    public void Quack(){
        Trk.Gobble();
    }
    public void Fly(){
        int I;
        for(I=0 ; I<5 ; I++)
            Trk.Fly();
    }
}
public class AdapterDesignPattern {
    public static void main(String args[]){
        Duck D1=new MallardDuck();
        D1.Quack();
        D1.Fly();
        System.out.println("==================");

        Turkey T1=new WildTurkey();
        T1.Gobble();
        T1.Fly();
        System.out.println("==================");

        AdapterForTurkey AFT=new AdapterForTurkey(T1);
        AFT.Quack();
        AFT.Fly();
        System.out.println("==================");
    }
}

//This is ADP Implemented With Object Adapter, Because Class Adapter is not Supported
//By Java Due to Multiple Inheritance
